In [1]:
from numpy import *
from pandas import *
from matplotlib.pyplot import *
from seaborn import *
from datetime import *
from sklearn.decomposition import PCA

import matplotlib
import time

In [2]:
# Settings
# data = 32*32
trainPath = 'C:/Users/Affinity/Documents/digits/trainingDigits/'
testPath = 'C:/Users/Affinity/Documents/digits/testDigits/'
numDigits_train = {0: 189, 1: 198, 2: 195, 3: 199, 4: 186, 5: 187, 6: 195, 7: 201, 8: 180, 9: 204}
numDigits_test = {0: 87, 1: 97, 2: 92, 3: 85, 4: 114, 5: 108, 6: 87, 7: 96, 8: 91, 9: 89}
nTrain=1934; nTest=946

## 1.1. Reading in txt data & preprocessing

#### Preprocessing train data

In [3]:
# preprocessing train data
ct=-1
trainData=zeros((nTrain,32**2))
trainLabel=zeros((nTrain,1))
for d in arange(10):
    currN = numDigits_train[d]
    for n in arange(currN):
        ct+=1
        temp=read_csv(trainPath+str(d)+ '_' +str(n)+ '.txt', header=None)
        temp1=[]
        for a in arange(32):
            temp1.extend(temp[0][a])
        trainData[ct]=temp1
        trainLabel[ct] = str(d)
print(trainData.shape)
print(trainLabel.shape)

(1934, 1024)
(1934, 1)


#### Preprocessing test data

In [4]:
# preprocessing test data
ct=-1
testData=zeros((nTest,32**2))
testLabel=zeros((nTest,1))
for d in arange(10):
    currN = numDigits_test[d]
    for n in arange(currN):
        ct+=1
        temp=read_csv(testPath+str(d)+ '_' +str(n)+ '.txt', header=None)
        temp1=[]
        for a in arange(32):
            temp1.extend(temp[0][a])
        testData[ct]=temp1
        testLabel[ct] = str(d)
print(testData.shape)
print(testLabel.shape)

(946, 1024)
(946, 1)


In [5]:
# checking data
info(trainData)

class:  ndarray
shape:  (1934, 1024)
strides:  (8192, 8)
itemsize:  8
aligned:  True
contiguous:  True
fortran:  False
data pointer: 0x18ad091e040
byteorder:  little
byteswap:  False
type: float64


In [6]:
# checking data
info(testData)

class:  ndarray
shape:  (946, 1024)
strides:  (8192, 8)
itemsize:  8
aligned:  True
contiguous:  True
fortran:  False
data pointer: 0x18ad18c4040
byteorder:  little
byteswap:  False
type: float64


In [7]:
info(trainLabel)
info(testLabel)

class:  ndarray
shape:  (1934, 1)
strides:  (8, 8)
itemsize:  8
aligned:  True
contiguous:  True
fortran:  True
data pointer: 0x18ac9f89dd0
byteorder:  little
byteswap:  False
type: float64
class:  ndarray
shape:  (946, 1)
strides:  (8, 8)
itemsize:  8
aligned:  True
contiguous:  True
fortran:  True
data pointer: 0x18acf1207e0
byteorder:  little
byteswap:  False
type: float64


## 1.2. Dimensionality reduction of data - PCA

In [8]:
nPC=70

In [9]:
pca=PCA(n_components=nPC)

In [10]:
pca.fit(trainData)

PCA(copy=True, iterated_power='auto', n_components=70, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)

In [11]:
pca_train=pca.transform(trainData)

In [12]:
print(pca.explained_variance_ratio_)
sum(pca.explained_variance_ratio_)

[0.10143927 0.09672018 0.08135553 0.05745237 0.04227788 0.03833902
 0.03271394 0.02779662 0.02493347 0.02249972 0.01976664 0.0168493
 0.01466366 0.01339683 0.0122538  0.0116096  0.01057918 0.01031074
 0.0090841  0.00884421 0.00780049 0.00775892 0.007097   0.00684513
 0.00585943 0.00553803 0.00511242 0.00477604 0.00459315 0.00450415
 0.00434546 0.00403067 0.00393917 0.00378237 0.00349449 0.00328494
 0.0031626  0.0030731  0.00299805 0.00282545 0.00277407 0.00273138
 0.00255856 0.00247755 0.00242456 0.00234015 0.00226825 0.00214205
 0.00209052 0.00202253 0.00198527 0.00193936 0.00191091 0.00187349
 0.00179048 0.0017324  0.00170411 0.00167593 0.00165635 0.00160749
 0.00158917 0.00157969 0.00154442 0.00151794 0.0015116  0.0014966
 0.00147061 0.00144559 0.00138456 0.00137685]


0.8043295662356361

#### Now, test data

In [13]:
pca=PCA(n_components=nPC)

In [14]:
pca.fit(testData)

PCA(copy=True, iterated_power='auto', n_components=70, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)

In [15]:
pca_test=pca.transform(testData)

In [16]:
print(pca.explained_variance_ratio_)
sum(pca.explained_variance_ratio_)

[0.10425934 0.09946362 0.08268231 0.05542169 0.04190473 0.03710207
 0.03282525 0.02725315 0.02457042 0.02262245 0.02112688 0.01938184
 0.0157538  0.01344201 0.01235616 0.01135315 0.010203   0.00970495
 0.00909071 0.00843887 0.00782551 0.00753967 0.007293   0.00664381
 0.00607012 0.00572598 0.00543527 0.00479729 0.00470543 0.00454933
 0.0042922  0.00409116 0.00386247 0.00378926 0.00356118 0.00341897
 0.00330992 0.00319956 0.0030644  0.00299209 0.00297006 0.00277272
 0.00265168 0.00254427 0.00251589 0.00239581 0.00232771 0.00225352
 0.00222474 0.00211165 0.00210709 0.0020721  0.00203086 0.00196586
 0.00192989 0.00191424 0.00185031 0.00182559 0.00178926 0.0017097
 0.00166027 0.00164338 0.00162233 0.00159338 0.00157242 0.00154515
 0.00150701 0.00147227 0.00146025 0.00145704]


0.8146194570982819

### Checking results of PCA

In [17]:
info(pca_train)
info(pca_test)

class:  ndarray
shape:  (1934, 70)
strides:  (560, 8)
itemsize:  8
aligned:  True
contiguous:  True
fortran:  False
data pointer: 0x18ad31b7040
byteorder:  little
byteswap:  False
type: float64
class:  ndarray
shape:  (946, 70)
strides:  (560, 8)
itemsize:  8
aligned:  True
contiguous:  True
fortran:  False
data pointer: 0x18acf4c0030
byteorder:  little
byteswap:  False
type: float64


## 1.3. Classification
### Methods: SVM, KNN
### Output: Number which the classifier predicted and whether it's correct, number of errors, ratio of errors

### 1.3.1. Applying SVM

In [18]:
from sklearn.svm import SVC
from sklearn.metrics import *

#### Hyperparameter Tuning

In [19]:
# Merging the train data with its label and then shuffling 

temp_train=concatenate((pca_train, trainLabel),axis=1)
temp_train=DataFrame(temp_train)
temp_train.head()

0         1         2         3         4         5         6   \
0  3.957719 -2.977937 -5.064481 -3.938011  3.023629  1.352358 -2.167631   
1  4.167249 -5.827844 -4.682440 -3.097088  0.950824  3.913542 -1.609677   
2  4.063555  0.594971 -5.064929 -3.325169  4.253401  2.245873 -2.259636   
3  3.391804 -2.206702 -4.786792 -3.978033  3.190582  3.111025 -1.115487   
4  5.001008 -4.268603 -5.167457 -2.972770 -0.233570  1.814039 -0.839027   

         7         8         9   ...        61        62        63        64  \
0 -1.108018  0.362550 -0.832895  ... -0.650721 -0.017875  0.351941  0.752845   
1 -1.027513  1.499876 -2.073185  ... -0.607848 -0.362568 -0.037630 -0.960731   
2 -2.770031 -1.535617  1.380120  ...  0.668961 -0.628098 -0.615505 -0.271280   
3 -1.527023  1.158968 -2.567453  ...  0.184095  0.123955 -0.121287  0.385933   
4  0.816726  0.161215 -0.226026  ... -0.775328  0.256517 -0.047150  0.465824   

         65        66        67        68        69   70  
0 -0.144586  0.265645 -0.252712  0.458389  1.329645  0.0  
1  0.206712  0.023548 -0.544110 -0.507964 -0.621242  0.0  
2 -0.559871  0.150355  0.738962 -0.413592  0.143290  0.0  
3  0.092392  0.349386 -0.072829  0.392378  1.024350  0.0  
4 -0.488518  0.190530  0.003666 -0.814790  0.178662  0.0  

[5 rows x 71 columns]

In [20]:
temp_train1=temp_train.reindex(np.random.permutation(temp_train.index))

In [21]:
temp_train1.head()

0         1         2         3         4         5         6   \
108  3.569173 -4.235911 -4.316881 -2.397469  1.140640  4.263675 -2.138478   
58   5.658332 -3.647845 -3.824929 -0.952831  0.061450  3.295226 -0.528169   
623 -2.136127 -1.106709  3.626609  2.635734  0.857538 -0.408993 -1.949389   
504 -2.369768 -1.940686  0.624999 -1.078621  1.067191  1.751128  5.093440   
750 -4.277883 -0.496092 -5.125371  1.118324 -1.699527 -0.511783  2.144268   

           7         8         9   ...        61        62        63  \
108 -0.288779  2.498928 -2.933593  ...  0.242638  0.041384 -0.164057   
58   0.111842  1.874021 -0.791805  ... -0.177441  0.362006 -0.980515   
623 -1.879087 -2.389602 -0.414401  ...  0.087354  0.393972 -0.177151   
504  3.289474 -1.344030 -2.647667  ... -0.124109  0.977059 -0.148795   
750 -1.984430 -3.944637 -1.982047  ... -0.256943  0.121739  0.258254   

           64        65        66        67        68        69   70  
108 -0.628812  0.946262  0.619322 -0.166951 -0.582294 -0.472305  0.0  
58  -0.598276  0.123541 -0.430738 -0.419491 -0.823697 -0.352696  0.0  
623  0.123555  0.247202  0.882959  0.101137 -0.305139 -0.174052  3.0  
504 -0.828257 -0.290366 -0.019984  0.908163 -0.251070  0.255731  2.0  
750  0.060033  0.417102  0.396645  0.055835 -0.632558 -0.236658  3.0  

[5 rows x 71 columns]

In [22]:
# Checking whether sorted correctly
unique(temp_train.iloc[int(temp_train1.iloc[0].name),:]==temp_train1.iloc[0,:])

array([ True])

In [23]:
# Splitting the data and label again
train1 = temp_train1.iloc[:1000,:-2]
train2 = temp_train1.iloc[1001:,:-2]
label1 = temp_train1.iloc[:1000,-1]
label2 = temp_train1.iloc[1001:,-1]

print(train1.shape, train2.shape, label1.shape, label2.shape)

(1000, 69) (933, 69) (1000,) (933,)


In [24]:
# Set the parameters by cross-validation:
# https://www.kaggle.com/udaysa/svm-with-scikit-learn-svm-with-parameter-tuning

In [44]:
# Grid search
best_score=0
for gamma in [0.001, 0.01, 0.1, 1, 10, 100]:
    for C in [0.001, 0.01, 0.1, 1, 10, 100]:
        # 매개변수의 각 조합에 대해 SVC를 훈련시킵니다.
        svm = SVC(gamma=gamma, C=C)
        svm.fit(train1,label1)
        # 테스트 세트로 SVC를 평가합니다
        score = svm.score(train2,label2)
        # 점수가 더 높으면 매개변수와 함께 기록
        if score>best_score:
            best_score=score
            best_parameters={'C':C, 'gamma':gamma}
print("최고 점수: {:.2f}".format(best_score))
print("최적 매개변수:",best_parameters)

최고 점수: 0.98
최적 매개변수: {'C': 10, 'gamma': 0.001}


#### SVM Classification

In [45]:
svc=SVC(**best_parameters)

In [46]:
svc.fit(pca_train,trainLabel)

C:\ProgramData\Anaconda3\envs\mypython\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(C=10, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=0.001, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [47]:
a=svc.predict(pca_test)
a.shape

(946,)

In [48]:
info(testLabel)

class:  ndarray
shape:  (946, 1)
strides:  (8, 8)
itemsize:  8
aligned:  True
contiguous:  True
fortran:  True
data pointer: 0x18acf1207e0
byteorder:  little
byteswap:  False
type: float64


In [49]:
errors=0
accuracyTable=zeros(nTest)
for t in arange(nTest):
    print("the classifier came back with: {0}, the real answer is: {1}" .format(int(a[t]), int(testLabel[t])))
    if (int(a[t])==int(testLabel[t])):
        accuracyTable[t]=1
    else:
        errors+=1
print("the total number of errors is: {}" .format(errors))
print("the total error rate is: {}" .format((errors/nTest)))

the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answer is: 0
the classifier came back with: 6, the real answe

In [50]:
print("훈련 세트 정확도: {:.2f}".format(svc.score(pca_train,trainLabel)))
print("테스트 세트 정확도: {:.2f}".format(svc.score(pca_test,testLabel)))

훈련 세트 정확도: 1.00
테스트 세트 정확도: 0.28


### 1.3.2. Applying k-NN
### Use the non-PCA version for k-NN

In [51]:
import operator

In [52]:
def classify0(inX, dataSet, labels, k):
    dataSetSize = dataSet.shape[0]
    diffMat = tile(inX, (dataSetSize,1))-dataSet
    sqDiffMat = diffMat**2
    sqDistances = sqDiffMat.sum(axis=1)
    distances = sqDistances**0.5
    sortedDistIndices = distances.argsort()
    classCount={}
    for i in range(k):
        votellabel = int(labels[sortedDistIndices[i]])
        classCount[votellabel] = classCount.get(votellabel,0)
    sortedClassCount = sorted(classCount.items(), key=operator.itemgetter(1),reverse=True)
    return sortedClassCount[0][0]

In [53]:
errors = 0
accuracyTable=zeros(nTest)
for kn in arange(testData.shape[0]):
    a=classify0(testData[kn], trainData, trainLabel,3)
    print("the classifier came back with: {0}, the real answer is: {1}" .format(int(a), int(testLabel[kn])))
    if a==int(testLabel[kn]):
        accuracyTable[kn]=1
    else:
            errors+=1
print("the total number of errors is: {}" .format(errors))
print("the total error rate is: {}" .format((errors/nTest)))

the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answe

the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answer is: 3
the classifier came back with: 3, the real answe

the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answer is: 4
the classifier came back with: 4, the real answe

the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answer is: 6
the classifier came back with: 6, the real answe

the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answer is: 7
the classifier came back with: 7, the real answe

the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answer is: 9
the classifier came back with: 9, the real answe

### Intermediate summary: KNN outperforms SVM (using PCA data)? Could it be that PCA per se lowers the accuracy? -> Try SVM without PCA

### 1.3.2. Performing SVM without PCA data

In [54]:
# Set the parameters by cross-validation:
# https://www.kaggle.com/udaysa/svm-with-scikit-learn-svm-with-parameter-tuning

In [55]:
from sklearn.utils import shuffle

In [56]:
# Merging the train data with its label and then shuffling 
temp_train=trainData + trainLabel
temp_train=concatenate((trainData, trainLabel),axis=1)
temp_train=DataFrame(temp_train)
temp_train.head()

0     1     2     3     4     5     6     7     8     9     ...  1015  \
0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0  ...   0.0   
3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
4   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

   1016  1017  1018  1019  1020  1021  1022  1023  1024  
0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 1025 columns]

In [57]:
temp_train1=temp_train.reindex(np.random.permutation(temp_train.index))

In [58]:
temp_train1.head()

0     1     2     3     4     5     6     7     8     9     ...  1015  \
1729   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1864   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
383    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1253   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1799   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   1.0   

      1016  1017  1018  1019  1020  1021  1022  1023  1024  
1729   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   8.0  
1864   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   9.0  
383    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0  
1253   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   6.0  
1799   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   9.0  

[5 rows x 1025 columns]

In [59]:
# Checking whether sorted correctly
unique(temp_train.iloc[int(temp_train1.iloc[0].name),:]==temp_train1.iloc[0,:])

array([ True])

In [60]:
# Splitting the data and label again
train1 = temp_train1.iloc[:1000,:-2]
train2 = temp_train1.iloc[1001:,:-2]
label1 = temp_train1.iloc[:1000,-1]
label2 = temp_train1.iloc[1001:,-1]

print(train1.shape, train2.shape, label1.shape, label2.shape)

(1000, 1023) (933, 1023) (1000,) (933,)


In [62]:
# Grid search
best_score=0
for gamma in [0.001, 0.01, 0.1, 1, 10, 100]:
    for C in [0.001, 0.01, 0.1, 1, 10, 100]:
        # 매개변수의 각 조합에 대해 SVC를 훈련시킵니다.
        svm = SVC(gamma=gamma, C=C)
        svm.fit(train1,label1)
        # 테스트 세트로 SVC를 평가합니다
        score = svm.score(train2,label2)
        # 점수가 더 높으면 매개변수와 함께 기록
        if score>best_score:
            best_score=score
            best_parameters={'C':C, 'gamma':gamma}
print("최고 점수: {:.2f}".format(best_score))
print("최적 매개변수:",best_parameters)

최고 점수: 0.98
최적 매개변수: {'C': 10, 'gamma': 0.01}


#### SVM Classification

In [63]:
svc=SVC(**best_parameters)

In [64]:
svc.fit(trainData,trainLabel)

C:\ProgramData\Anaconda3\envs\mypython\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(C=10, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=0.01, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [65]:
a=svc.predict(testData)
a.shape

(946,)

In [66]:
info(testLabel)

class:  ndarray
shape:  (946, 1)
strides:  (8, 8)
itemsize:  8
aligned:  True
contiguous:  True
fortran:  True
data pointer: 0x18acf1207e0
byteorder:  little
byteswap:  False
type: float64


In [67]:
errors=0
accuracyTable=zeros(nTest)
for t in arange(nTest):
    print("the classifier came back with: {0}, the real answer is: {1}" .format(int(a[t]), int(testLabel[t])))
    if (int(a[t])==int(testLabel[t])):
        accuracyTable[t]=1
    else:
        errors+=1
print("the total number of errors is: {}" .format(errors))
print("the total error rate is: {}" .format((errors/nTest)))

the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answer is: 0
the classifier came back with: 0, the real answe

In [68]:
print("훈련 세트 정확도: {:.2f}".format(svc.score(trainData,trainLabel)))
print("테스트 세트 정확도: {:.2f}".format(svc.score(testData,testLabel)))

훈련 세트 정확도: 1.00
테스트 세트 정확도: 0.99


### Conclusion: SVM performs better on data without PCA than data after PCA (0.3% VS. 0.88%). And SVM without PCA outperforms KNN also (8 errors vs. 12 errors)